modules/withRouter.js

```diff
@@ -1,31 +1,43 @@
+import invariant from 'invariant'
 import React from 'react'
 import hoistStatics from 'hoist-non-react-statics'
 import { routerShape } from './PropTypes'
-import warning from './routerWarning'
-
 
 function getDisplayName(WrappedComponent) {
   return WrappedComponent.displayName || WrappedComponent.name || 'Component'
 }
 
 export default function withRouter(WrappedComponent, options) {
-  const { withRef } = options || {}
+  const withRef = options && options.withRef
 
   const WithRouter = React.createClass({
     contextTypes: { router: routerShape },
     propTypes: { router: routerShape },
 
     getWrappedInstance() {
-      warning(withRef, 'To access the wrappedInstance you must provide { withRef: true } as the second argument of the withRouter call')
-      return this.wrappedComponent
+      invariant(
+        withRef,
+        'To access the wrapped instance, you need to specify ' +
+        '`{ withRef: true }` as the second argument of the withRouter() call.'
+      )
+
+      return this.wrappedInstance
     },
 
     render() {
-      const { router, ...props } = this.props
-
-      if (withRef) props.ref = component =>this.wrappedComponent = component
-
-      return <WrappedComponent {...props} router={router || this.context.router} />
+      const router = this.props.router || this.context.router
+
+      if (withRef) {
+        return (
+          <WrappedComponent
+            {...this.props}
+            ref={(c) => { this.wrappedInstance = c }}
+            router={router}
+          />
+        )
+      }
+
+      return <WrappedComponent {...this.props} router={router} />
```

Comment:
I'd rather not branch around a return statement. As a general rule of thumb, that makes code harder to reason around because the exit point is no longer static.